setup repo & prep v0.1.0

riastrad opened the pull request:
This PR adds all of the code & stands up the important parts of the repo config. Here's a quick summary:

* adds all of the initial working code & tests
* sets up initial package.json for npm (though it won't be released until after we merge this)
* sets up Travis for the repo

**still to do**
- [x] update README.md
- [x] add CODEOWNERS

Commit:
add all code

Commit:
npm package setup

Commit:
fix up some paths and names

Commit:
eslint --fix .

Commit:
remove coverage crut && add gitignore

Commit:
add travis support

Commit:
no need for different npm creds

Commit:
README & CHANGELOG

Commit:
ease coverage restriction on branches

Commit:
fix broken test

Commit:
add CODEOWNERS

riastrad commented:
@samanpwbb @tristen this is ready for a formal review whenever either of you have a cycle to spare. Once we're 👍 I'll merge & run out a formal release of the new package.

tristen commented on index.js:
```diff
@@ -0,0 +1,22 @@
+const stitchTiles = require('./src/stitch-tiles');
+
+
+function makeGetQuilt(coordinates, zoom, length = DEFAULT_LENGTH, decode = true) {
+  return (style) => {
```
Curious why this isn't the other way around. Wouldn't it be more common to initialize with ` { style, token }` and then call it repeatedly with `coordinates`, and `zoom`?

riastrad replied:
Currently I'm just focused on keeping this as it's currently being used downstream, which is why the pattern is structured this way. But I'm not opposed to inverting the `make/get` relationship, as it does make more sense the way you've outlined.

tristen replied:
Obvs your call here but is this something you would like to do instead? It's a pretty breaking change if it's introduced later on.

riastrad replied:
I did a quick review of the downstream implementations and I think shifting this structure would cause us some non-trivial refactoring for this use case. My gut reaction is to defer as much downstream refactoring as possible, and accept this pattern as is for now.

If we decide to use this library in more places in the future (and find ourselves frustrated by the pattern), then the breaking change would be more worth the effort.

tristen commented on index.js:
```diff
@@ -0,0 +1,22 @@
+function makeGetQuilt(coordinates, zoom, length = DEFAULT_LENGTH, decode = true) {
+  return (style) => {
+    const pixelCoords = getPixelCoordinates(coordinates, zoom);
+    const finalImage = getSiblingTilesAndOffsets(pixelCoords, zoom, length)
```
nit but this could just return? The named variable is clearer though.

riastrad replied:
Yeah, I purposefully opted for clarity over concision here 👍

tristen commented on index.js:
```diff
@@ -0,0 +1,22 @@
+  return (style) => {
+    const pixelCoords = getPixelCoordinates(coordinates, zoom);
+    const finalImage = getSiblingTilesAndOffsets(pixelCoords, zoom, length)
+      .then((imageIngredients) => { return getAllTiles(imageIngredients, style); })
```
nit but these chains can just be written out as 

```suggestion
      .then(imageIngredients => getAllTiles(imageIngredients, style)); 
```

riastrad replied:
Great point, changed in 53b2f84 👍

tristen commented on src/get-all-static-tiles.js:
```diff
@@ -0,0 +1,22 @@
+function getAllTiles(tileImgProperties, { style, accessToken }) {
+  const tiles = tileImgProperties.tiles;
+  const promises = [];
+  tiles.forEach((tile) => {
```
```suggestion
  tiles.forEach(tile => {
```

riastrad replied:
Hmm, I get a linting error from our Mapbox config when I do this 🤔

https://github.com/mapbox/eslint-config-mapbox/blob/master/base.js#L15

tristen replied:
Oh! I guess we require parens. Let's just close these suggestions 🙈

tristen commented on src/get-all-static-tiles.js:
```diff
@@ -0,0 +1,22 @@
+    promises.push(getStaticTile(tileUrl, { x: px, y: py }));
+  });
+
+  return Promise.all(promises).catch((e) => {
```
```suggestion
  return Promise.all(promises).catch(e => {
```

riastrad replied:
`tiles.map` is a great suggestion! 🙌

tristen commented on src/get-all-static-tiles.js:
```diff
@@ -0,0 +1,22 @@
+    promises.push(getStaticTile(tileUrl, { x: px, y: py }));
+  });
+
+  return Promise.all(promises).catch((e) => {
```
`promises` could also just be a `tiles.map`

riastrad replied:
`tiles.map` is a great suggestion! 🙌

tristen commented on src/get-sibling-tiles-and-offsets.js:
```diff
@@ -0,0 +1,60 @@
+    zoom: centerIndex.zoom
+  };
+
+  return Promise.resolve(imgProperties).catch((e) => { throw new Error(e); });
```
Would this ever catch?

riastrad replied:
🤔 I don't think so, actually. I'll remove it and see what happens.

tristen commented on src/get-sibling-tiles-and-offsets.js:
```diff
@@ -0,0 +1,60 @@
+  // in order to make the final image
+  for (let row = topLeftIndex.row; row <= bottomRightIndex.row; row++) {
+    if (row < 0) continue;
+    if (row >= maxTiles) break;
```
Would this just be `>` to factor the `maxTiles`?

riastrad replied:
Yep, good catch.

riastrad replied:
Hmmm, even though tests are passing in my latest commit, I'm seeing some weird false-positive behavior for the get-sibling-tiles-and-offsets tests now: 

![Screen Shot 2020-04-29 at 12 03 17 PM](https://user-images.githubusercontent.com/16272584/80618852-ed633380-8a11-11ea-9655-435ef1b7aea8.png)
[link](https://travis-ci.com/github/mapbox/quilt/jobs/325085201)

riastrad replied:
I'll need to dig into this a bit more to understand why this change is causing that behavior, but I suspect it's primarily to do with how the test is written since the change lines did not modify how `zoom` values are handled at all.

riastrad replied:
Ah, yep - bad fixture. Fixed in 4d7622b

tristen commented on src/stitch-tiles.js:
```diff
@@ -0,0 +1,29 @@
+        return resolve(
+          {
+            decoded: decode ? PNG.sync.read(img) : null,
+            original: img.toString('base64')
```
Thoughts on just returning the base64 encoded string or binary? Personally, I feel like consumers should create the png if they need to. quilt can keep pretty lean by not including `pngjs` as a dependency.

riastrad replied:
Ooh, I like where your head's at. I _would_ like to keep quilt as light as possible, so I can definitely get on board with pushing responsibility for decoding to downstream consumers. 👍

riastrad replied:
Switching to returning just the raw buffer in https://github.com/mapbox/quilt/pull/1/commits/9abde5f45017c0ac14c494e351ad572794241092

tristen commented on utils/check-token.js:
```diff
@@ -0,0 +1,12 @@
+
+function checkToken(token) {
+
+  if (!token || !token.match(/^(sk|pk)\.([^/]*)$/)) {
```
should `tk` be included?

riastrad replied:
Not sure home frequently a `tk` token is needed for this use case, but can definitely add one 👍

riastrad replied:
added in c8d4937

daniel-j-h commented:
Hello lovely people :hugs: - I had the use for exactly this a while ago and just used rio back then

https://github.com/mapbox/rasterio/issues/1846#issuecomment-565531023

How's this different and why not just using the rio one-liner?

Commit:
temp tokens are still valid tokens

riastrad commented:
👋 hello @daniel-j-h, thanks for flagging that slick rasterio one-liner! I had no idea rasterio could deal with slippy maps that easily.

I'm not super familiar with using rasterio, but here's the primary difference I can name off the top of my head: `@mapbox/quilt` is written in Node specifically so that we can use it downstream with some of our internal debugging tools that are also written in Node, and we'd prefer not to mix python & node within these tools.

There may be other good reasons, but I would need to be more familiar with the behavior and uses of `rio calc` before I could say definitively.

Commit:
clean map of tiles

Commit:
remove needless catch & included max

Commit:
clip those curls

Commit:
stitch tiles returns raw buffer

allows us to remove pngjs dependency

Commit:
fix index test

Commit:
fix incorrect fixture

Commit:
pin mapnik version

stop gap to ensure node v10 compatibility until I can land https://github.com/mapbox/quilt/issues/2

riastrad commented:
@tristen @samanpwbb this is ready for a final review.

tristen reviewed (approved):
🧶 🖼️
